### ISSUE TITLE
alpha vs lr_policy for the Adam solver

### ISSUE THREAD

**Issue by jeff-regier** [pluskid/Mocha.jl#107]
The documentation states that `alpha` is a parameter to the Adam solver. However the Adam code sets `alpha` based on the `lr_policy` (and not necessarily a `LRPolicy.Fixed`) rather than an explicitly specification of `alpha`. So I think the doc and the code aren't consistent.

Specifying an `lr_policy` rather than a fixed `alpha` is certainly more flexible, but is that much flexibility desirable? I was under the impression that the point of Adam was that the solver picked the learning rate. Any guidelines for when using a non-constant learning rate improves the convergence rate? Thanks.

**Comment by benmoran** [comment]
The initial version did indeed use a fixed learning rate, called `alpha` in line with the Adam paper.  And it's true that Adam is able to adapt the learning rate itself to some extent.  

However, the adaptation is more for adjusting individual parameters relative to each other, rather than for setting the overall rate - in fact the "alpha"/"learning_rate" parameter is more or less an upper bound to the amount any parameter can change at a single step.  So depending on the scaling of the parameters, it can still be useful to start with a higher rate and anneal it downwards as you get closer to the optimum.

In fact they do use this in the Adam paper as well - in their Logistic Regression experiment 6.1 they mention that the plots are made with alpha_t = alpha/sqrt_t.  That looks similar but not identical to to LRPolicy.Inv in Mocha. I've also heard other people suggest that a variable rate is beneficial with Adam.

On the other hand,  in Dirk Kingma's code here https://github.com/dpkingma/nips14-ssl/blob/1fc7aec899ca429ff26a1d10f747e5ad6bcd17b2/gpulearn_yz_x.py he seems to be using a fixed 0.003 learning rate.

So YMMV, it seems like it's still trial and error to find the best settings for your particular problems, unfortunately.  I would be glad to hear what you end up using.

**Comment by jeff-regier** [comment]
Hi Ben, Thanks for your response. I guess for now I'll do a grid search over many possible settings of `lr_policy`. I might try switching to a second-order optimization method eventually, e.g. http://arxiv.org/pdf/1509.02866v1.pdf .
